SimExLite/PMICalculators/SimpleScatteringPMICalculator.py

```diff
@@ -0,0 +1,11147 @@
+
+        self.f_s2e_setup()
+
+    def f_s2e_setup(self):
+        self.g_s2e["setup"] = dict()
+        self.g_s2e["sys"] = dict()
+        self.g_s2e["setup"]["num_digits"] = 7
+        self.g_s2e["steps"] = 100
+        self.g_s2e["maxZ"] = 100
+
+    ##############################################################################
+
+    def f_save_info(self):
+        pmi_file = self.g_s2e["setup"]["pmi_out"]
+
+        grp = "/info"
+        ###    print pmi_file , grp
+        xfp = h5py.File(pmi_file, "a")
+        try:
+            grp_hist_parent = xfp.create_group(grp)
+        except:
+            1
+
+        #    xfp[ grp + '/Sq_free' ] = g_dbase['Sq_free']
+        xfp.close()
+
+    ##############################################################################
+
+    def f_init_random(self):
+        random.seed(self.g_s2e["id"])
+
+    ##############################################################################
+
+    def f_dbase_Zq2id(self, a_Z, a_q):
+        return (a_Z * (a_Z + 1)) // 2 - 1 * 1 + a_q
+
+    ##############################################################################
+
+    def f_dbase_setup(self):
+        # print '   Update database!!!'
+        xdbase = load_ff_database()
+
+        g_dbase = dict()
+
+        #   q               ->   sin(theta/2)/lambda
+        #   au, exp(i*q*r)  ->   1/Angstrom, exp( 2*pi*q*r)
+        g_dbase["halfQ"] = xdbase[:, 0] = xdbase[:, 0] / (
+            2.0 * numpy.pi * 0.529177206 * 2.0
+        )
+
+        maxZ = self.g_s2e["maxZ"]  # 99
+        numQ = len(g_dbase["halfQ"])
+        g_dbase["ff"] = numpy.zeros((self.f_dbase_Zq2id(maxZ, maxZ) + 1, numQ))
+        ii = 0
+        for ZZ in range(1, maxZ + 1):
+            for qq in range(ZZ + 1):
+                g_dbase["ff"][ii, :] = xdbase[:, ZZ] * (ZZ - qq) / (ZZ * 1.0)
+                ii = ii + 1
+
+        g_dbase["Sq_halfQ"] = g_dbase["halfQ"]
+        g_dbase["Sq_bound"] = numpy.zeros((numQ,))
+        g_dbase["Sq_free"] = numpy.zeros((numQ,))
+
+        self.g_dbase = g_dbase
+
+    #    g_dbase['ph_sigma'] ;
+
+    ##    g_dbase['halfQ'] = numpy.array( [ 0 , 1 , 2 ] ) ;
+    ##    maxZ = g_s2e['maxZ']  #99
+    ##    g_dbase['ff'] = numpy.zeros( ( f_dbase_Zq2id( maxZ , maxZ ) + 1 , len( g_dbase['halfQ'] ) ) )
+    ##    ii = 0
+    ##    for ZZ in range( 1 , maxZ+1 ) :
+    ##        for qq in range( ZZ+1 ) :
+    ##            g_dbase['ff'][ ii , : ] = numpy.array( [ 1.0 , 0.5 , 0.25 ] ) * ( ZZ - qq ) ;
+    #####            print ZZ,  qq, ii
+    #####            print  g_dbase['ff'][ ii , : ]
+    ##            ii = ii + 1
+    ##    g_dbase['Sq_halfQ'] = numpy.array( [ 0 , 0 , 0 ] ) ;
+    ##    g_dbase['Sq_bound'] = numpy.array( [ 0 , 0 , 0 ] ) ;
+    ##    g_dbase['Sq_free']  = numpy.array( [ 0 , 0 , 0 ] ) ;
+    ###    g_dbase['ph_sigma'] ;
+
+    def f_load_snp_content(self, a_fp, a_snp):
+        dbase_root = (
+            "/data/snp_" + str(a_snp).zfill(self.g_s2e["setup"]["num_digits"]) + "/"
```

Comment:
This is not portable and will fail on Windows. If there's ever a chance this will run on a Wndows environment, `os.path.join` is better.